backend/msgpack/conn.go

```diff
@@ -0,0 +1,164 @@
+// THE SOFTWARE.
+
+package msgpack
+
+import (
+	"errors"
+	"math"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/m3db/m3x/clock"
+)
+
+const (
+	tcpProtocol = "tcp"
+)
+
+var (
+	errNoActiveConnection = errors.New("no active connection")
+)
+
+type connectWithLockFn func() error
+type writeWithLockFn func([]byte) error
+
+// connection is a persistent connection that retries establishing
+// connection with exponential backoff if the connection goes down.
+type connection struct {
+	sync.Mutex
+
+	addr         string
+	opts         ConnectionOptions
+	connTimeout  time.Duration
+	writeTimeout time.Duration
+	nowFn        clock.NowFn
+
+	conn        *net.TCPConn
+	numFailures int
+	threshold   int
+
+	connectWithLockFn connectWithLockFn
+	writeWithLockFn   writeWithLockFn
+}
+
+// newConnection creates a new connection.
+func newConnection(addr string, opts ConnectionOptions) *connection {
+	c := &connection{
+		addr:         addr,
+		opts:         opts,
+		connTimeout:  opts.ConnectionTimeout(),
+		writeTimeout: opts.WriteTimeout(),
+		nowFn:        opts.ClockOptions().NowFn(),
+		threshold:    opts.InitReconnectThreshold(),
+	}
+	c.connectWithLockFn = c.connectWithLock
+	c.writeWithLockFn = c.writeWithLock
+
+	c.Lock()
+	c.connectWithLockFn()
+	c.Unlock()
+
+	return c
+}
+
+// Write sends data onto the connection, and attempts to re-establish
+// connection if the connection is down.
+func (c *connection) Write(data []byte) error {
+	c.Lock()
+	if c.conn == nil {
+		c.numFailures++
+		if !c.checkReconnectWithLock() {
+			c.Unlock()
+			return errNoActiveConnection
+		}
+	}
+	writeErr := c.writeWithLockFn(data)
+	if writeErr == nil {
+		c.Unlock()
+		return nil
+	}
+	if err := c.connectWithLockFn(); err == nil {
+		if writeErr = c.writeWithLockFn(data); writeErr == nil {
+			c.Unlock()
+			return nil
+		}
+	}
+	c.numFailures++
+	c.closeWithLock()
+	c.Unlock()
+	return writeErr
+}
+
+func (c *connection) Close() {
+	c.Lock()
+	c.closeWithLock()
+	c.Unlock()
+}
+
+func (c *connection) connectWithLock() error {
+	conn, err := net.DialTimeout(tcpProtocol, c.addr, c.connTimeout)
+	if err != nil {
+		return err
+	}
+	tcpConn := conn.(*net.TCPConn)
+	tcpConn.SetKeepAlive(c.opts.ConnectionKeepAlive())
+	if c.conn != nil {
+		c.conn.Close()
+	}
+	c.conn = tcpConn
+	return nil
+}
+
+func (c *connection) checkReconnectWithLock() bool {
+	// If we haven't accumulated enough failures to warrant another reconnect
+	// empty, simply returning false without reconnecting.
+	if c.numFailures <= c.threshold {
+		return false
+	}
+	if err := c.connectWithLockFn(); err == nil {
+		c.resetWithLock()
+		return true
+	}
+	if maxThreshold := c.opts.MaxReconnectThreshold(); c.threshold < maxThreshold {
+		newThreshold := c.threshold * c.opts.ReconnectThresholdMultiplier()
+		c.threshold = int(math.Min(float64(newThreshold), float64(maxThreshold)))
+	}
+	return false
+}
+
+func (c *connection) writeWithLock(data []byte) error {
```

Comment:
Yeah I'm leaving the lock as is. I could use a RWMutex instead of a mutex and acquire a read lock here, but given that only one goroutine is going to perform the write, it's unclear whether using a RWMutex will bring any material perf benefit.